src/main/java/com/atlassian/db/replica/api/AuroraLsnReplicaConsistency.java

```diff
@@ -0,0 +1,99 @@
+package com.atlassian.db.replica.api;
+
+import com.atlassian.db.replica.internal.MonotonicMemoryCache;
+import com.atlassian.db.replica.spi.Cache;
+import com.atlassian.db.replica.spi.ReplicaConsistency;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * It remembers last write LSN (log sequence number) and compares it with LSN for subsequent reads to detect inconsistencies.
+ * If it cannot remember the LSN of last write, pessimistically assumes it's going to be inconsistent.
+ *
+ * @see <a href="https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-global-database-monitoring.html">Monitoring Aurora PostgreSQL-based Aurora global databases</a>
+ * <p>
+ * It's an PostgreSQL-based Aurora database engine specific implementation.
+ */
+public class AuroraLsnReplicaConsistency implements ReplicaConsistency {
+
+    private final Cache<Long> lastWrite;
+
+    public static class Builder {
+        private Cache<Long> lastWrite = new MonotonicMemoryCache<>();
+
+        /**
+         * @param lastWrite remembers last write
+         */
+        public Builder cacheLastWrite(Cache<Long> lastWrite) {
+            this.lastWrite = lastWrite;
+            return this;
+        }
+
+        /**
+         * @return consistency assuming that LSN (log sequence number) is greater or equal to LSN for last write (if known)
+         */
+        public AuroraLsnReplicaConsistency build() {
+            return new AuroraLsnReplicaConsistency(lastWrite);
+        }
+    }
+
+    private AuroraLsnReplicaConsistency(Cache<Long> lastWrite) {
+        this.lastWrite = lastWrite;
+    }
+
+    @Override
+    public void write(Connection main) {
+        try {
+            lastWrite.put(queryMainDbLsn(main));
+        } catch (Exception e) {
+            lastWrite.reset();
+        }
+    }
+
+    @Override
+    public boolean isConsistent(Supplier<Connection> replica) {
+        Optional<Long> maybeLastWrite = lastWrite.get();
+        if (!maybeLastWrite.isPresent()) {
+            return false;
+        }
+
+        long lastRefresh;
+        try {
+            lastRefresh = queryReplicaDbLsn(replica.get());
+        } catch (Exception e) {
+            return false;
+        }
+
+        return lastRefresh >= maybeLastWrite.get();
+    }
+
+    /**
+     * @return LSN (log sequence number) for the main database (the highest LSN)
+     */
+    private long queryMainDbLsn(Connection connection) throws Exception {
+        return queryLsn(connection, "SELECT MAX(durable_lsn) AS lsn FROM aurora_global_db_instance_status();");
```

Comment:
We are checking it for the current connection/replica. What if max(LSN) for one replica will be up to date and max LSN for the current one will not. Do we have a test to cover it?

Reply:
Ok, I see it's for the main database 🤣 .

Reply:
AFAIR in Postgres, it was more lightweight as we could select LSN instead of aggregate it.

Reply:
Right, using "aurora" function is more convenient. In the spare time we can also add version with "postgres" lsn and compare them. "Postgres" solution has this downside that is highly dependent on postgres version - functions names are different for different postgres versions. I guess "aurora" api will be more stable.

Reply:
We removed the Postgres one from the API to [avoid dependency on Postgres JDBC](https://github.com/atlassian-labs/db-replica/issues/9). We were using it to parse LSN.

Reply:
So how to avoid dependencies when implementing LSN ?
LSN is dependent on database

Reply:
1. Aurora provided LSN as long already, so no need to depend on a parser.
2. We can write our own Postgres `LSN to long` parser.
3. We can create another project `postgres-db-replica` with the extra dependency on Postgres JDBC. It will not block `db-replica` re-use for projects that don't need Postgres JDBC dependency.
4. We can implement PostgresLSN consistency in the product (ReplicaConsistency is an SPI).

Reply:
Ok, if i understood correctly "aurora" solution is better because we avoid adding dependency on Postgres JDBC.